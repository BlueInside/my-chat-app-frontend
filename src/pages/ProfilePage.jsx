import { useLoaderData } from 'react-router-dom';
import UserProfileHeader from '../components/UserProfileHeader';
import { useAuth } from '../utils/AuthContext';
import UserInfoDisplay from '../components/UserInfoDisplay';
import styled from 'styled-components';
import { Link } from 'react-router-dom';

const Container = styled.div`
  display: flex;
  padding: 20px;
  background-color: #f9f9f9;
  border-radius: 8px;
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
  max-width: 600px;
  margin: 0 auto;
  width: 100%;
  flex-direction: column;
  height: 100%;
  justify-content: center;
`;

const EditButton = styled(Link)`
  display: inline-block;
  margin-top: 20px;
  padding: 10px 20px;
  background-color: #007bff;
  color: white;
  text-decoration: none;
  border-radius: 4px;
  width: fit-content;
  &:hover {
    background-color: #0056b3;
  }
`;

function ProfilePage() {
  const user = useLoaderData();
  const { user: currentUser } = useAuth();
  return (
    <Container>
      <UserProfileHeader avatarUrl={user.avatarUrl} username={user.username} />
      <UserInfoDisplay user={user} />
      {currentUser.id === user._id && (
        <EditButton to={'edit'}>Edit Profile</EditButton>
      )}
    </Container>
  );
}

export default ProfilePage;
