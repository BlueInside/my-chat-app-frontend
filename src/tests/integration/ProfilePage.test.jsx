import { render, screen } from '@testing-library/react';
import { RouterProvider, createMemoryRouter } from 'react-router-dom';
import { afterEach, describe, expect, it, vi } from 'vitest';
import ProfilePage from '../../pages/ProfilePage';

const mockUseAuth = vi.fn();
vi.mock('../../utils/AuthContext', () => ({
  useAuth: () => mockUseAuth(),
}));

const mockUserData = {
  _id: 'user1',
  username: 'Karol',
  fullName: 'Karol Pulawski',
  bio: 'My very interesting bio',
  avatarUrl: '#',
  dateOfBirth: '1997 05 18',
};

const router = createMemoryRouter(
  [
    {
      path: '/',
      element: <ProfilePage />,
      loader: () => mockUserData,
    },
    { path: 'edit', element: <div>Edit profile page</div> },
  ],
  { initialEntries: ['/'] }
);

describe('ProfilePage Component', async () => {
  afterEach(() => {
    vi.clearAllMocks();
  });

  it('Renders Profile page properly', () => {
    mockUseAuth.mockReturnValue({ user: { id: 'user1' } });
    const { container } = render(<RouterProvider router={router} />);

    expect(container).toMatchSnapshot();
  });

  it(`Should display edit button when current user id is same as profile param id`, async () => {
    mockUseAuth.mockReturnValue({ user: { id: 'user1' } });
    render(<RouterProvider router={router} />);

    const editProfileButton = screen.getByRole('link', {
      name: /edit profile/i,
    });

    expect(editProfileButton).toBeInTheDocument();
  });

  it(`Shouldn't display edit button when current user id is different than profile param id`, async () => {
    mockUseAuth.mockReturnValue({ user: { id: 'user2' } });

    render(<RouterProvider router={router} />);

    const editProfileButton = screen.queryByRole('link', {
      name: /edit profile/i,
    });

    expect(editProfileButton).not.toBeInTheDocument();
  });
});
