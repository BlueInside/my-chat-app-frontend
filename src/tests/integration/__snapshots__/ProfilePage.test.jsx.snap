// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProfilePage Component > Renders Profile page properly 1`] = `
<div>
  <div
    class="sc-iAEzLF ikFkrg"
  >
    <div
      class="sc-beqWNU ceBucd"
    >
      <img
        alt="avatar"
        class="sc-gueXAH ivoMqK"
        src="#"
      />
      <p
        class="sc-dmqGRC cBaAVu"
      >
        Karol
      </p>
    </div>
    <div
      class="sc-hLseSp cuzUdV"
    >
      <div
        class="sc-eDDOjk bBRgum"
      >
        <span
          class="sc-jTrOVX ha-dZIX"
        >
          Username:
        </span>
        <span
          class="sc-gLDymS gUZinN"
        >
          Karol
        </span>
      </div>
      <div
        class="sc-eDDOjk bBRgum"
      >
        <span
          class="sc-jTrOVX ha-dZIX"
        >
          Full Name:
        </span>
        <span
          class="sc-gLDymS gUZinN"
        >
          Karol Pulawski
        </span>
      </div>
      <div
        class="sc-eDDOjk bBRgum"
      >
        <span
          class="sc-jTrOVX ha-dZIX"
        >
          Bio:
        </span>
        <span
          class="sc-gLDymS gUZinN"
        >
          My very interesting bio
        </span>
      </div>
      <div
        class="sc-eDDOjk bBRgum"
      >
        <span
          class="sc-jTrOVX ha-dZIX"
        >
          Role:
        </span>
        <span
          class="sc-gLDymS gUZinN"
        />
      </div>
      <div
        class="sc-eDDOjk bBRgum"
      >
        <span
          class="sc-jTrOVX ha-dZIX"
        >
          Last Active:
        </span>
        <span
          class="sc-gLDymS gUZinN"
        >
          Invalid Date
        </span>
      </div>
      <div
        class="sc-eDDOjk bBRgum"
      >
        <span
          class="sc-jTrOVX ha-dZIX"
        >
          Account Created:
        </span>
        <span
          class="sc-gLDymS gUZinN"
        >
          Invalid Date
        </span>
      </div>
    </div>
    <a
      class="sc-fsQjcA ldzKiz"
      href="/edit"
    >
      Edit Profile
    </a>
  </div>
</div>
`;
